Android/app/src/main/java/app/intra/ServerChooserFragment.java

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2018 Jigsaw Operations LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package app.intra;
+
+import android.app.Dialog;
+import android.os.Bundle;
+import android.support.v7.app.AlertDialog;
+import android.support.v7.preference.PreferenceDialogFragmentCompat;
+import android.text.Editable;
+import android.text.TextWatcher;
+import android.view.KeyEvent;
+import android.view.View;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.RadioGroup;
+import android.widget.TextView;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+import app.intra.util.Untemplate;
+
+/**
+ * User interface for a the server URL selection.
+ */
+
+public class ServerChooserFragment extends PreferenceDialogFragmentCompat
+    implements RadioGroup.OnCheckedChangeListener, TextWatcher, EditText.OnEditorActionListener {
+    private String newUrl = null;
+    private RadioGroup buttons = null;
+    private EditText text = null;
+
+    public static ServerChooserFragment newInstance(String key) {
+        final ServerChooserFragment fragment = new ServerChooserFragment();
+        final Bundle bundle = new Bundle(1);
+        bundle.putString(ARG_KEY, key);
+        fragment.setArguments(bundle);
+        return fragment;
+    }
+
+    @Override
+    public void onCheckedChanged(RadioGroup group, int checkedId) {
+        if (checkedId == R.id.pref_server_google) {
+            newUrl = getResources().getString(R.string.url0);
+            setValid(true);
```

Comment:
I personally like to set the UI based on state.  E.g. update newURL, then call an updateUI() function, or have a setter of some sort for newURL. Makes it easier to maintain the UI logic.

Reply:
OK, done